Merge with scrobbler

rafaelgomesxyz opened the pull request:
Fixes #19 

Still need to merge the actual scrobblers, but most of it is done.

Commit:
Merge models

Commit:
Merge Trakt APIs

Commit:
Merge services

Commit:
Merge popup

Commit:
Merge options

Commit:
Move streaming service files to a separate folder

Commit:
Implement logic to separate scrobbler from sync

Commit:
Merge common

Commit:
Merge Netflix scrobbler

Commit:
Merge Amazon Prime scrobbler

MrMamen commented on src/_locales/en/messages.json:
```diff
@@ -15,7 +15,7 @@
 	},
 	"aboutMessage": {
-		"message": "Sync the history from your favorite streaming services to Trakt.tv!"
+		"message": "Scrobble / sync the history from your favorite streaming services to Trakt.tv!"
```
Could we simplify and just say: "Send your activity from your favorite streaming services to Trakt.tv"?

rafaelgomesxyz replied:
Yup

MrMamen commented on src/components/UtsLeftRight.tsx:
```diff
@@ -5,7 +5,7 @@ interface UtsLeftRightProps {
 	className?: string;
 	left: React.ReactElement;
-	right?: React.ReactElement;
+	right?: React.ReactElement | null;
```
Does it have to be optional if it can be null?

rafaelgomesxyz replied:
You're right, it's better to leave it just as optional, since this allows us to omit the property entirely if we don't need it.

MrMamen commented on src/modules/background/background.ts:
```diff
@@ -1,33 +1,75 @@
-	url: string;
+export type MessageRequest =
+	| {
+			action:
```
Probably why a lot of people don't like TypeScript, but I prefer defining every single one of these as Interfaces. It will be a bit more overhead, but I think think it looks cleaner.

rafaelgomesxyz replied:
I agree. It also makes it easier in the future if we need to add a property to an action.

MrMamen commented on src/_locales/pt_BR/messages.json:
```diff
@@ -66,6 +66,15 @@
+		"message": "Não foi possível realizar o scrobble."
+	},
+	"disableScrobblingDescription": {
+		"message": "With this option enabled, the extension will not automatically scrobble what you are watching. This is useful if you only want to use the history sync."
```
Remember these. ;-)

rafaelgomesxyz replied:
Hehe, coincidentally I've actually added an eslint plugin for translation files to the hbo-go branch, and I modified it to check for duplicate messages (https://github.com/godaddy/eslint-plugin-i18n-json/pull/33), which can show cases like this as errors, so we never forget to translate messages again. \o/

Commit:
Merge HBO Go scrobbler

MrMamen committed:
Update messages.json

Commit:
Fix types

Commit:
Improve event typings

Commit:
Merge branch 'merge-scrobbler' of https://github.com/trakt-tools/universal-trakt-sync into merge-scrobbler

rafaelgomesxyz commented on src/services/Events.ts:
```diff
@@ -5,70 +5,121 @@ import { StreamingServiceId } from '../streaming-services/streaming-services';
 import { Errors } from './Errors';
 import { RequestException } from './Requests';
-
-export enum Events {
```
I don't see the benefit of using an enum here, is there one? Using a conditional type makes it much easier to do generic typings. If you think an enum is best, I can revert.

Commit:
Update about message

Commit:
Rename everything from sync to scrobbler

Commit:
Fix HBO Go parser

Commit:
Reduce duplicate code for content scripts

Commit:
Fix lint-staged

Commit:
Add templates to make it easier to add new services

Commit:
Rename 'services' folder to 'common' to avoid confusion with 'streaming-services'

Commit:
Reduce duplicate code in Events

Commit:
Fix styles

rafaelgomesxyz commented:
Alright, think I'm done here.

rafaelgomesxyz commented:
The only bug I'm experiencing is that sometimes the popup doesn't load, which appears to be the same bug that affects traktflix (https://github.com/tegon/traktflix/issues/100). But after about 3 times of clicking the button, it loads. I'm trying to see what's causing this.

**Edit:** Actually, doesn't look like it's that issue, but I was sure this also happened in traktflix.

MrMamen reviewed (changes requested):
I get an error when toggling: Show browser notifications. (Failed to save option.)

MrMamen commented:
> I get an error when toggling: Show browser notifications. (Failed to save option.)

Same with Netflix. Not able to disable it.

rafaelgomesxyz commented:
> I get an error when toggling: Show browser notifications. (Failed to save option.)

For notifications, it's because I forgot to add the `notifications` permission to the manifest, but I don't know why you can't disable Netflix. I'll test with a clean profile.

MrMamen commented:
> > I get an error when toggling: Show browser notifications. (Failed to save option.)
> 
> For notifications, it's because I forgot to add the `notifications` permission to the manifest, but I don't know why you can't disable Netflix. I'll test with a clean profile.

Probably related but I'm not able to scrobble from Netflix either. I did try to clean all cache & data, but same problem.

rafaelgomesxyz commented:
> Probably related but I'm not able to scrobble from Netflix either. I did try to clean all cache & data, but same problem.

Ok, I can reproduce it on Chromium.

**Edit:** This is a bizarre error. It looks like the extension still thinks that Netflix is a required permission. I tried completely wiping the extension from my Chromium folder and it still thinks that.

Commit:
Add missing notifications permission

Commit:
Fix bug that does not scrobble Netflix or HBO Go on Chromium

rafaelgomesxyz commented:
Ok, I fixed the scrobble bug.

As for not being able to disable Netflix, I think this is because we're loading the extension as unpacked, so it considers all permissions as required. I don't think this should happen with an actual extension coming from the store.

Commit:
Fix Netflix API params event

MrMamen commented:
> Ok, I fixed the scrobble bug.
> 
> 
> 
> As for not being able to disable Netflix, I think this is because we're loading the extension as unpacked, so it considers all permissions as required. I don't think this should happen with an actual extension coming from the store.

But why are the others toggable?

rafaelgomesxyz commented:
> > Ok, I fixed the scrobble bug.
> > As for not being able to disable Netflix, I think this is because we're loading the extension as unpacked, so it considers all permissions as required. I don't think this should happen with an actual extension coming from the store.
> 
> But why are the others toggable?

They're not toggable for me. Probably because I erased the extension and added it again.

MrMamen commented:
> > > Ok, I fixed the scrobble bug.
> > > As for not being able to disable Netflix, I think this is because we're loading the extension as unpacked, so it considers all permissions as required. I don't think this should happen with an actual extension coming from the store.
> > 
> > 
> > But why are the others toggable?
> 
> They're not toggable for me. Probably because I erased the extension and added it again.

Just for the record. The services with scrobbling are not toggable for me, but the ones with just history sync are.
